src/Cache/FileHandler.php

```diff
@@ -41,21 +43,86 @@ public function read(): ?CacheInterface
             return null;
         }
 
-        $content = file_get_contents($this->file);
-
-        try {
-            $cache = Cache::fromJson($content);
-        } catch (\InvalidArgumentException $exception) {
+        $handle = fopen($this->file, 'r');
+        if (false === $handle) {
             return null;
         }
 
+        $cache = $this->readFromHandle($handle);
+        $this->fileLastModification = $this->getFileLastUpdate($handle);
+        var_dump("\n".getmypid().' !!! FRS !!! READ '.$this->fileLastModification);
+        fclose($handle);
+
         return $cache;
     }
 
     public function write(CacheInterface $cache): void
     {
-        $content = $cache->toJson();
+        $this->ensureFileIsWriteable();
+
+        $handle = fopen($this->file, 'r+');
+        if (false === $handle) {
+            return;
+        }
+
+        $actualLastModification = $this->getFileLastUpdate($handle);
+
+        var_dump("\n".getmypid().' !!! FRS !!! DECISION ???');
+        var_dump([
+            'property' => $this->fileLastModification,
+            'os' => $actualLastModification,
+            'should backfill' => $this->fileLastModification < $actualLastModification,
+        ]);
+
+        if ($this->fileLastModification < $actualLastModification) {
+            flock($handle, LOCK_EX);
+
+            $oldCache = $this->readFromHandle($handle);
+            rewind($handle);
+
+            if ($oldCache && method_exists($cache, 'backfillHashes')) {
+                var_dump("\n".getmypid().' !!! FRS !!! BACKFILL');
+                $cache->backfillHashes($oldCache);
+            }
+        }
+        var_dump("\n".getmypid().' !!! FRS !!! SAVE FILE');
+        ftruncate($handle, 0);
+        fwrite($handle, $cache->toJson());
```

Comment:
I wonder about putting `@` to silence warning and check `!== false` for every method here... any better suggestion?